src/sentry/snuba/query_subscription_consumer.py

```diff
@@ -0,0 +1,205 @@
+from json import loads
+
+import jsonschema
+from confluent_kafka import Consumer, KafkaException, TopicPartition
+from django.conf import settings
+
+from sentry.snuba.json_schemas import SUBSCRIPTION_PAYLOAD_VERSIONS, SUBSCRIPTION_WRAPPER_SCHEMA
+from sentry.snuba.models import QuerySubscription
+from sentry.utils import metrics
+
+logger = logging.getLogger(__name__)
+
+
+subscriber_registry = {}
+
+
+def register_subscriber(subscriber_key):
+    def inner(func):
+        if subscriber_key in subscriber_registry:
+            raise Exception("Handler already registered for %s" % subscriber_key)
+        subscriber_registry[subscriber_key] = func
+        return func
+
+    return inner
+
+
+class InvalidMessageError(Exception):
+    pass
+
+
+class InvalidSchemaError(InvalidMessageError):
+    pass
+
+
+class QuerySubscriptionConsumer(object):
+    """
+    A Kafka consumer that processes query subscription update messages. Each message has
+    a related subscription id and the latest values related to the subscribed query.
+    These values are passed along to a callback associated with the subscription.
+    """
+
+    def __init__(
+        self, group_id, topic=None, commit_batch_size=100, initial_offset_reset="earliest"
+    ):
+        self.group_id = group_id
+        if not topic:
+            topic = settings.KAFKA_SNUBA_QUERY_SUBSCRIPTIONS
+        self.topic = topic
+        cluster_name = settings.KAFKA_TOPICS[topic]["cluster"]
+        self.bootstrap_servers = settings.KAFKA_CLUSTERS[cluster_name]["bootstrap.servers"]
+        self.commit_batch_size = commit_batch_size
+        self.initial_offset_reset = initial_offset_reset
+
+    def run(self):
+        logger.debug("Starting snuba query subscriber")
+
+        conf = {
+            "bootstrap.servers": self.bootstrap_servers,
+            "group.id": self.group_id,
+            "session.timeout.ms": 6000,
+            "auto.offset.reset": self.initial_offset_reset,
+            "enable.auto.commit": "false",
+            "enable.auto.offset.store": "false",
+            "enable.partition.eof": "false",
+            "default.topic.config": {"auto.offset.reset": self.initial_offset_reset},
+        }
+
+        def on_revoke(consumer, partitions):
+            commit_offsets()
+
+        def commit_offsets():
+            if offsets:
+                to_commit = [
+                    TopicPartition(self.topic, partition, offset)
+                    for partition, offset in offsets.items()
+                ]
+                consumer.commit(offsets=to_commit)
+                offsets.clear()
+
+        consumer = Consumer(conf)
+        consumer.subscribe([self.topic], on_revoke=on_revoke)
+
+        offsets = {}
+
+        def shutdown():
+            logger.debug("Committing offsets and closing consumer")
+            commit_offsets()
+            consumer.close()
+
+        def handler(signum, frame):
+            shutdown()
+
+        signal.signal(signal.SIGINT, handler)
```

Comment:
I would register the SIGINT handler in the command file so you decouple the concerns of running the consumer and the logic of the consumer, thus allowing the consumer to be used in a different context than a sentry command.